use `towncrier` to handle changelog entries

zacharyburnett opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->

<!-- If this PR closes a GitHub issue, reference it here by its number -->
related to #8670 

<!-- describe the changes comprising this PR here -->
using `towncrier` to handle changelog entries will
- drastically reduce (or even eliminate) branch conflicts in `CHANGES.rst` we currently experience with PRs
- keep the sections in `CHANGES.rst` consistent and eliminate duplicate sections
- allow automatic links to be generated in the changelog entries

[<img width="1152" alt="image" src="https://github.com/user-attachments/assets/24873697-fe5d-4102-83a4-32b1bf2995dd">](https://jwst-pipeline--8671.org.readthedocs.build/en/8671/jwst/changes.html)

`towncrier` expects "news fragment" files (text files in the `changes/` directory with filenames in the format `<PR#>.<changetype>.rst`, i.e. for this PR it would be `changes/8671.docs.rst`). See docs at https://towncrier.readthedocs.io/en/latest/tutorial.html#creating-news-fragments

when ready to make a release, run `towncrier build` to ingest the news fragments and generate a changelog section in `CHANGES.rst` with all the new change log entries for that release (this clears the `changes/` directory of all news fragment files). This step should either be done before making a release, or could probably be added to a GitHub workflow triggered on release (to insert a commit and remake the tag).

After merging this PR the `Release Process` wiki page will need to be updated to include a step to run `towncrier build` instead of manually editing `CHANGES.rst`

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

zacharyburnett commented on .github/pull_request_template.md:
```diff
@@ -9,7 +9,7 @@ Closes #
 
 **Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
-- [ ] added entry in `CHANGES.rst` within the relevant release section
+- [ ] write a news fragment in `changes/` with the format `<#>.<type>.rst`, i.e. for pull request #8670 `changes/8670.feature.rst`. Types are `feature`, `bugfix`, `doc`, `removal`, and `misc`.
```
instructions in the pull request checklist for making a news fragment

zacharyburnett commented on .github/pull_request_template.md:
```diff
@@ -9,7 +9,7 @@ Closes #
 
 **Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
-- [ ] added entry in `CHANGES.rst` within the relevant release section
+- [ ] write a [news fragment in `changes/` with the format `<pr#>.<feature|bugfix|doc|removal|misc>.rst`](https://towncrier.readthedocs.io/en/latest/tutorial.html#creating-news-fragments), i.e. `changes/8671.misc.rst`
```
instructions in the pull request checklist for making a news fragment

zacharyburnett commented:
as @braingram pointed out in tagup this morning, using `towncrier` also has the advantage of reducing merge conflicts in `CHANGES.rst` when multiple PRs change the same changelog section, as well as preventing the edge case of a long-standing PR putting its changelog entry into an old version section

braingram commented:
Would this require adding a subdirectory and `pyproject.toml` entry for each possible heading? The current changelog has many headings:
https://github.com/spacetelescope/jwst/blob/1a45e3480c6e7d6c103a188c04ce4041084fc644/CHANGES.rst?plain=1#L4-L5
and looking at astropy's use of towncrier it has a subdirectory for each section:
https://github.com/astropy/astropy/tree/main/docs/changes/config
and a corresponding entry in [pyproject.toml](https://github.com/astropy/astropy/blob/9a7002fcf4de719116a6bf3b9752fcb41f5a44f2/pyproject.toml#L476)

Also, is it required to manually run `towncrier` for each build or is this something the CI can handle?

braingram commented:
> If we want to keep the current headings, then yes. However, I also wanted to discuss whether it would be more useful for the user to condense the headings to the defaults here; do the current module-based headings make the changelog more or less confusing?

I don't have an informed opinion here so I'll keep quiet :)

zacharyburnett commented:
> > If we want to keep the current headings, then yes. However, I also wanted to discuss whether it would be more useful for the user to condense the headings to the defaults here; do the current module-based headings make the changelog more or less confusing?
> 
> I don't have an informed opinion here so I'll keep quiet :)

calling @nden; should we keep the current changelog headings? I can add them here

tapastro commented:
There are use cases for the changelog during development - it's convenient to have a single location to read all updates since the last release. Is there another option besides only building the changelog on release under towncrier?

My 2c would be that the subheadings are very useful and should remain, fwiw.

zacharyburnett commented:
> There are use cases for the changelog during development - it's convenient to have a single location to read all updates since the last release. Is there another option besides only building the changelog on release under towncrier?

The downside, however, is that the developer must manually update the changelog with every release, and changelog entries might be accidentally inserted into old changelog sections. The `changelog/` directory will still have all of the updates since the last release (the files in it are cleared by the `towncrier` action when creating a new changelog section)

zacharyburnett commented:
also @tapastro, you can manually run `towncrier build --draft` at any time to output the resulting changelog section to console output

zacharyburnett commented:
current output of `towncrier build --draft`:
```
❯ towncrier build --draft      
Loading template...
Finding news fragments...
Rendering news fragments...
Draft only -- nothing has been written.
What is seen below is what would be written.

1.15.2.dev187+gf2475922c (2024-08-19)
=====================================

Cube_build
----------

- Removed direct setting of the ``self.skip`` attribute from within the step itself. (#8600)


Master_background
-----------------

- Either of ``"background"`` or ``"bkg"`` in slit name now defines the slit as a background slit, instead of ``"bkg"`` only. (#8600)


Outlier_detection
-----------------

- Fixed failures due to a missing ``wcs.array_shape`` attribute when the ``outlier_detection`` step was run standalone using e.g. ``strun`` (#8645)


Scripts
-------

- Removed many non-working and out-dated scripts. Including many scripts that were replaced by ``strun``. (#8619)


Stpipe
------

- Removed setting of the `self.skip` attribute in the `record_step_status()` function; added a `query_step_status()` function to use as an alternative to checking `self.skip`. (#8600)


Tweakreg
--------

- Removed direct setting of the ``self.skip`` attribute from within the step itself. (#8600)
- Updated requirement for ``tweakwcs`` to version ``0.8.8`` which fixes a crash in the ``tweakreg`` step due to a ``MalformedPolygonError`` exception being raised by the ``spherical_geometry`` package for some data sets. (spacetelescope/tweakwcs#205) (#8657)
```

zacharyburnett commented:
new changelog after rebasing and adding links:
```
❯ towncrier build --draft
Loading template...
Finding news fragments...
Rendering news fragments...
Draft only -- nothing has been written.
What is seen below is what would be written.

1.15.2.dev187+gf2475922c (2024-08-19)
=====================================

Align_refs
----------

- Compute alignment shifts from the first integration of the science exposure
  only. (`#8643 <https://github.com/spacetelescope/jwst/8643>`_)


Ami_average
-----------

- Fix error in step spec that prevents step creation. (`#8677
  <https://github.com/spacetelescope/jwst/8677>`_)


Assign_wcs
----------

- Moved `update_s_region_imaging`, `update_s_region_keyword`, and
  `wcs_from_footprints` into stcal. (`#8624
  <https://github.com/spacetelescope/jwst/8624>`_)


Cube_build
----------

- Removed direct setting of the ``self.skip`` attribute from within the step
  itself. (`#8600 <https://github.com/spacetelescope/jwst/8600>`_)


Emicorr
-------

- Fixed a bug where MIRI EMI correction step would return NaNs when it was
  unable to compute a correction. (`#8675
  <https://github.com/spacetelescope/jwst/8675>`_)


General
-------

- Remove the unused ``stsci.image`` dependency. (`#8663
  <https://github.com/spacetelescope/jwst/8663>`_)
- Update required stcal version to 1.8.0. (`#8706
  <https://github.com/spacetelescope/jwst/8706>`_)


Master_background
-----------------

- Either of ``"background"`` or ``"bkg"`` in slit name now defines the slit as
  a background slit, instead of ``"bkg"`` only. (`#8600
  <https://github.com/spacetelescope/jwst/8600>`_)


Outlier_detection
-----------------

- Fixed failures due to a missing ``wcs.array_shape`` attribute when the
  ``outlier_detection`` step was run standalone using e.g. ``strun`` (`#8645
  <https://github.com/spacetelescope/jwst/8645>`_)


Resample_spec
-------------

- - Modified the output NIRSpec spectral WCS to sample the input data linearly
  in sky
    coordinates, rather than slit coordinates, in order to conserve spectral
    flux in default reductions.

  - Updated handling for the ``pixel_scale_ratio`` parameter to apply only to
  the
    spatial dimension, to match the sense of the parameter application to the
    documented intent, and to conserve spectral flux when applied.

  - Implemented handling for the ``pixel_scale`` parameter, which was
  previously
    ignored for spectral resampling.

  - Fixed a bug resulting in incorrect output slit coordinates for NIRSpec
  moving
    targets in the ``calwebb_spec3`` pipeline. (`#8596
  <https://github.com/spacetelescope/jwst/8596>`_)
- Separate ``resample_spec`` step parameters from ``resample`` step parameters
  so that the spectral resampling step only exposes parameters that are
  appropriate for spectral data. (`#8622
  <https://github.com/spacetelescope/jwst/8622>`_)


Scripts
-------

- Removed many non-working and out-dated scripts. Including many scripts that
  were replaced by ``strun``. (`#8619
  <https://github.com/spacetelescope/jwst/8619>`_)


Set_telescope_pointing
----------------------

- Refactored separate modes into submodules instead of inheriting from a base
  class. Moved non-JWST-specific code to stcal. (`#8613
  <https://github.com/spacetelescope/jwst/8613>`_)
- replace usage of ``copy_arrays=True`` with ``memmap=False`` (`#8660
  <https://github.com/spacetelescope/jwst/8660>`_)


Stpipe
------

- Removed setting of the `self.skip` attribute in the `record_step_status()`
  function; added a `query_step_status()` function to use as an alternative to
  checking `self.skip`. (`#8600
  <https://github.com/spacetelescope/jwst/8600>`_)


Tso_photometry
--------------

- Replaced photutils.aperture do_photometry with photutils.ApertureStats to
  remove NaNs when performing TSO photometry for non SUB64 WLP8 pupil data. For
  SUB64 WLP8 changed summing data in the aperture to use np.nansum to ignore
  NaNs. (`#8672 <https://github.com/spacetelescope/jwst/8672>`_)


Tweakreg
--------

- Removed direct setting of the ``self.skip`` attribute from within the step
  itself. (`#8600 <https://github.com/spacetelescope/jwst/8600>`_)
- Moved all realignment methods to stcal. (`#8624
  <https://github.com/spacetelescope/jwst/8624>`_)
- Updated requirement for ``tweakwcs`` to version ``0.8.8`` which fixes a crash
  in the ``tweakreg`` step due to a ``MalformedPolygonError`` exception being
  raised by the ``spherical_geometry`` package for some data sets.
  (spacetelescope/tweakwcs#205) (`#8657
  <https://github.com/spacetelescope/jwst/8657>`_)
```

zacharyburnett commented:
this will play nicely with #8716

zacharyburnett commented:
additionally, after merging this PR #8670 will also have to be updated

zacharyburnett commented:
Right now, the subsections in the changelog are ordered alphabetically. However, I see some utility in ordering the changelog sections *by step* in the pipeline, which users will be familiar with. We can specify a manual order by using a TOML array to define change types, instead of TOML tables: 
https://towncrier.readthedocs.io/en/stable/configuration.html#use-a-toml-array-defined-order

zacharyburnett commented:
```rst
1.15.2.dev195+g3c053d10a (2024-08-27)
=====================================

General
-------

- Remove the unused ``stsci.image`` dependency. (`#8663
  <https://github.com/spacetelescope/jwst/8663>`_)
- Update required stcal version to 1.8.0. (`#8706
  <https://github.com/spacetelescope/jwst/8706>`_)
- Increase minimum required stpipe. (`#8713
  <https://github.com/spacetelescope/jwst/8713>`_)


Documentation
-------------

- use ``towncrier`` to handle change log entries (`#8671
  <https://github.com/spacetelescope/jwst/8671>`_)


``stpipe``
----------

- Removed setting of the `self.skip` attribute in the `record_step_status()`
  function; added a `query_step_status()` function to use as an alternative to
  checking `self.skip`. (`#8600
  <https://github.com/spacetelescope/jwst/8600>`_)


Associations
------------

- Restored slit name to level 3 product names for NIRSpec BOTS and background
  fixed slit targets. (`#8699 <https://github.com/spacetelescope/jwst/8699>`_)


Scripts
-------

- Removed many non-working and out-dated scripts. Including many scripts that
  were replaced by ``strun``. (`#8619
  <https://github.com/spacetelescope/jwst/8619>`_)


``set_telescope_pointing``
--------------------------

- Refactored separate modes into submodules instead of inheriting from a base
  class. Moved non-JWST-specific code to stcal. (`#8613
  <https://github.com/spacetelescope/jwst/8613>`_)
- replace usage of ``copy_arrays=True`` with ``memmap=False`` (`#8660
  <https://github.com/spacetelescope/jwst/8660>`_)


``align_refs`` (``coron3``)
---------------------------

- Compute alignment shifts from the first integration of the science exposure
  only. (`#8643 <https://github.com/spacetelescope/jwst/8643>`_)


``ami_analyze`` / ``ami_normalize`` / ``ami_average`` (``ami3``)
----------------------------------------------------------------

- Fix error in step spec that prevents step creation. (`#8677
  <https://github.com/spacetelescope/jwst/8677>`_)


``assign_wcs`` (``image2``, ``spec2``)
--------------------------------------

- Moved `update_s_region_imaging`, `update_s_region_keyword`, and
  `wcs_from_footprints` into stcal. (`#8624
  <https://github.com/spacetelescope/jwst/8624>`_)


``cube_build`` (``spec2`` IFU, ``spec3``)
-----------------------------------------

- Removed direct setting of the ``self.skip`` attribute from within the step
  itself. (`#8600 <https://github.com/spacetelescope/jwst/8600>`_)


``emicorr`` (``detector1`` MIR)
-------------------------------

- Fixed a bug where MIRI EMI correction step would return NaNs when it was
  unable to compute a correction. (`#8675
  <https://github.com/spacetelescope/jwst/8675>`_)


``master_background`` (``spec2`` MOS, ``spec3``)
------------------------------------------------

- Either of ``"background"`` or ``"bkg"`` in slit name now defines the slit as
  a background slit, instead of ``"bkg"`` only. (`#8600
  <https://github.com/spacetelescope/jwst/8600>`_)


``outlier_detection`` (``image3``, ``tso3``, ``spec3``, ``coron3``)
-------------------------------------------------------------------

- Fixed failures due to a missing ``wcs.array_shape`` attribute when the
  ``outlier_detection`` step was run standalone using e.g. ``strun`` (`#8645
  <https://github.com/spacetelescope/jwst/8645>`_)


``resample`` (``image2``, ``image3``, ``coron3``)
-------------------------------------------------

- Fixed a typo in ``load_custom_wcs`` from ``array_shape`` to ``pixel_shape``
  and changed to use values in the top-level ASDF structure if the values in
  the WCS are ``None``. (`#8698
  <https://github.com/spacetelescope/jwst/8698>`_)


``resample_spec`` (``spec2`` MOS, ``spec3``)
--------------------------------------------

- - Modified the output NIRSpec spectral WCS to sample the input data linearly
  in sky
    coordinates, rather than slit coordinates, in order to conserve spectral
    flux in default reductions.

  - Updated handling for the ``pixel_scale_ratio`` parameter to apply only to
  the
    spatial dimension, to match the sense of the parameter application to the
    documented intent, and to conserve spectral flux when applied.

  - Implemented handling for the ``pixel_scale`` parameter, which was
  previously
    ignored for spectral resampling.

  - Fixed a bug resulting in incorrect output slit coordinates for NIRSpec
  moving
    targets in the ``calwebb_spec3`` pipeline. (`#8596
  <https://github.com/spacetelescope/jwst/8596>`_)
- Separate ``resample_spec`` step parameters from ``resample`` step parameters
  so that the spectral resampling step only exposes parameters that are
  appropriate for spectral data. (`#8622
  <https://github.com/spacetelescope/jwst/8622>`_)


``tso_photometry`` (``tso3``)
-----------------------------

- Replaced photutils.aperture do_photometry with photutils.ApertureStats to
  remove NaNs when performing TSO photometry for non SUB64 WLP8 pupil data. For
  SUB64 WLP8 changed summing data in the aperture to use np.nansum to ignore
  NaNs. (`#8672 <https://github.com/spacetelescope/jwst/8672>`_)


``tweakreg`` (``image3``)
-------------------------

- Removed direct setting of the ``self.skip`` attribute from within the step
  itself. (`#8600 <https://github.com/spacetelescope/jwst/8600>`_)
- Moved all realignment methods to stcal. (`#8624
  <https://github.com/spacetelescope/jwst/8624>`_)
- Updated requirement for ``tweakwcs`` to version ``0.8.8`` which fixes a crash
  in the ``tweakreg`` step due to a ``MalformedPolygonError`` exception being
  raised by the ``spherical_geometry`` package for some data sets.
  (spacetelescope/tweakwcs#205) (`#8657
  <https://github.com/spacetelescope/jwst/8657>`_)
```

zacharyburnett commented on .github/pull_request_template.md:
```diff
@@ -9,7 +9,79 @@ Closes #
 
 **Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
-- [ ] added entry in `CHANGES.rst` within the relevant release section
+- [ ] for a public change, added a [towncrier news fragment](https://towncrier.readthedocs.io/en/stable/tutorial.html#creating-news-fragments) <details><summary>`changes/<PR#>.<changetype>.rst`</summary>
```
**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
 - [ ] for a public change, added a [towncrier news fragment](https://towncrier.readthedocs.io/en/stable/tutorial.html#creating-news-fragments) <details><summary>`changes/<PR#>.<changetype>.rst`</summary>
     - ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
     - ``changes/<PR#>.docs.rst``
     - ``changes/<PR#>.stpipe.rst``
     - ``changes/<PR#>.datamodels.rst``
     - ``changes/<PR#>.scripts.rst``
     - ``changes/<PR#>.fits_generator.rst``
     - ``changes/<PR#>.set_telescope_pointing.rst``

     ## steps
     - ``changes/<PR#>.align_refs.rst``
     - ``changes/<PR#>.ami.rst``
     - ``changes/<PR#>.assign_mtwcs.rst``
     - ``changes/<PR#>.assign_wcs.rst``
     - ``changes/<PR#>.background.rst``
     - ``changes/<PR#>.barshadow.rst``
     - ``changes/<PR#>.charge_migration.rst``
     - ``changes/<PR#>.combine_1d.rst``
     - ``changes/<PR#>.cube_build.rst``
     - ``changes/<PR#>.cube_skymatch.rst``
     - ``changes/<PR#>.dark_current.rst``
     - ``changes/<PR#>.dq_init.rst``
     - ``changes/<PR#>.emicorr.rst``
     - ``changes/<PR#>.engdb_tools.rst``
     - ``changes/<PR#>.exp_to_source.rst``
     - ``changes/<PR#>.extract_1d.rst``
     - ``changes/<PR#>.extract_2d.rst``
     - ``changes/<PR#>.firstframe.rst``
     - ``changes/<PR#>.flatfield.rst``
     - ``changes/<PR#>.fringe.rst``
     - ``changes/<PR#>.gain_scale.rst``
     - ``changes/<PR#>.group_scale.rst``
     - ``changes/<PR#>.guider_cds.rst``
     - ``changes/<PR#>.imprint.rst``
     - ``changes/<PR#>.ipc.rst``
     - ``changes/<PR#>.jump.rst``
     - ``changes/<PR#>.klip.rst``
     - ``changes/<PR#>.lastframe.rst``
     - ``changes/<PR#>.linearity.rst``
     - ``changes/<PR#>.master_background.rst``
     - ``changes/<PR#>.mrs_imatch.rst``
     - ``changes/<PR#>.msaflagopen.rst``
     - ``changes/<PR#>.nsclean.rst``
     - ``changes/<PR#>.outlier_detection.rst``
     - ``changes/<PR#>.pathloss.rst``
     - ``changes/<PR#>.persistence.rst``
     - ``changes/<PR#>.photom.rst``
     - ``changes/<PR#>.pixel_replace.rst``
     - ``changes/<PR#>.ramp_fitting.rst``
     - ``changes/<PR#>.refpix.rst``
     - ``changes/<PR#>.resample.rst``
     - ``changes/<PR#>.resample_spec.rst``
     - ``changes/<PR#>.reset.rst``
     - ``changes/<PR#>.residual_fringe.rst``
     - ``changes/<PR#>.rscd.rst``
     - ``changes/<PR#>.saturation.rst``
     - ``changes/<PR#>.skymatch.rst``
     - ``changes/<PR#>.source_catalog.rst``
     - ``changes/<PR#>.spectral_leak.rst``
     - ``changes/<PR#>.srctype.rst``
     - ``changes/<PR#>.stack_refs.rst``
     - ``changes/<PR#>.straylight.rst``
     - ``changes/<PR#>.superbias.rst``
     - ``changes/<PR#>.timeconversion.rst``
     - ``changes/<PR#>.transforms.rst``
     - ``changes/<PR#>.tso_photometry.rst``
     - ``changes/<PR#>.tweakreg.rst``
     - ``changes/<PR#>.wavecorr.rst``
     - ``changes/<PR#>.wfs_combine.rst``
     - ``changes/<PR#>.wfss_contam.rst``
     - ``changes/<PR#>.white_light.rst``
     </details>
 - [ ] updated or added relevant tests
 - [ ] updated relevant documentation
 - [ ] added relevant milestone

zacharyburnett commented on .github/pull_request_template.md:
```diff
@@ -9,7 +9,79 @@ Closes #
+    - ``changes/<PR#>.jump.rst``
+    - ``changes/<PR#>.klip.rst``
+    - ``changes/<PR#>.lastframe.rst``
+    - ``changes/<PR#>.linearity.rst``
+    - ``changes/<PR#>.master_background.rst``
+    - ``changes/<PR#>.mrs_imatch.rst``
+    - ``changes/<PR#>.msaflagopen.rst``
+    - ``changes/<PR#>.nsclean.rst``
+    - ``changes/<PR#>.outlier_detection.rst``
+    - ``changes/<PR#>.pathloss.rst``
+    - ``changes/<PR#>.persistence.rst``
+    - ``changes/<PR#>.photom.rst``
+    - ``changes/<PR#>.pixel_replace.rst``
+    - ``changes/<PR#>.ramp_fitting.rst``
+    - ``changes/<PR#>.refpix.rst``
+    - ``changes/<PR#>.resample.rst``
+    - ``changes/<PR#>.resample_spec.rst``
+    - ``changes/<PR#>.reset.rst``
+    - ``changes/<PR#>.residual_fringe.rst``
+    - ``changes/<PR#>.rscd.rst``
+    - ``changes/<PR#>.saturation.rst``
+    - ``changes/<PR#>.skymatch.rst``
+    - ``changes/<PR#>.source_catalog.rst``
+    - ``changes/<PR#>.spectral_leak.rst``
+    - ``changes/<PR#>.srctype.rst``
+    - ``changes/<PR#>.stack_refs.rst``
+    - ``changes/<PR#>.straylight.rst``
+    - ``changes/<PR#>.superbias.rst``
+    - ``changes/<PR#>.timeconversion.rst``
+    - ``changes/<PR#>.transforms.rst``
+    - ``changes/<PR#>.tso_photometry.rst``
+    - ``changes/<PR#>.tweakreg.rst``
+    - ``changes/<PR#>.wavecorr.rst``
+    - ``changes/<PR#>.wfs_combine.rst``
+    - ``changes/<PR#>.wfss_contam.rst``
+    - ``changes/<PR#>.white_light.rst``
+    </details>
 - [ ] updated or added relevant tests
 - [ ] updated relevant documentation
 - [ ] added relevant milestone
```
**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
 - [ ] for a public change, added a towncrier news fragment in `changes/` <details><summary>`echo "changed something" > changes/<PR#>.<changetype>.rst`</summary>

     - ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
     - ``changes/<PR#>.docs.rst``
     - ``changes/<PR#>.stpipe.rst``
     - ``changes/<PR#>.datamodels.rst``
     - ``changes/<PR#>.scripts.rst``
     - ``changes/<PR#>.fits_generator.rst``
     - ``changes/<PR#>.set_telescope_pointing.rst``

     ## steps
     - ``changes/<PR#>.align_refs.rst``
     - ``changes/<PR#>.ami.rst``
     - ``changes/<PR#>.assign_mtwcs.rst``
     - ``changes/<PR#>.assign_wcs.rst``
     - ``changes/<PR#>.background.rst``
     - ``changes/<PR#>.barshadow.rst``
     - ``changes/<PR#>.charge_migration.rst``
     - ``changes/<PR#>.combine_1d.rst``
     - ``changes/<PR#>.cube_build.rst``
     - ``changes/<PR#>.cube_skymatch.rst``
     - ``changes/<PR#>.dark_current.rst``
     - ``changes/<PR#>.dq_init.rst``
     - ``changes/<PR#>.emicorr.rst``
     - ``changes/<PR#>.engdb_tools.rst``
     - ``changes/<PR#>.exp_to_source.rst``
     - ``changes/<PR#>.extract_1d.rst``
     - ``changes/<PR#>.extract_2d.rst``
     - ``changes/<PR#>.firstframe.rst``
     - ``changes/<PR#>.flatfield.rst``
     - ``changes/<PR#>.fringe.rst``
     - ``changes/<PR#>.gain_scale.rst``
     - ``changes/<PR#>.group_scale.rst``
     - ``changes/<PR#>.guider_cds.rst``
     - ``changes/<PR#>.imprint.rst``
     - ``changes/<PR#>.ipc.rst``
     - ``changes/<PR#>.jump.rst``
     - ``changes/<PR#>.klip.rst``
     - ``changes/<PR#>.lastframe.rst``
     - ``changes/<PR#>.linearity.rst``
     - ``changes/<PR#>.master_background.rst``
     - ``changes/<PR#>.mrs_imatch.rst``
     - ``changes/<PR#>.msaflagopen.rst``
     - ``changes/<PR#>.nsclean.rst``
     - ``changes/<PR#>.outlier_detection.rst``
     - ``changes/<PR#>.pathloss.rst``
     - ``changes/<PR#>.persistence.rst``
     - ``changes/<PR#>.photom.rst``
     - ``changes/<PR#>.pixel_replace.rst``
     - ``changes/<PR#>.ramp_fitting.rst``
     - ``changes/<PR#>.refpix.rst``
     - ``changes/<PR#>.resample.rst``
     - ``changes/<PR#>.resample_spec.rst``
     - ``changes/<PR#>.reset.rst``
     - ``changes/<PR#>.residual_fringe.rst``
     - ``changes/<PR#>.rscd.rst``
     - ``changes/<PR#>.saturation.rst``
     - ``changes/<PR#>.skymatch.rst``
     - ``changes/<PR#>.source_catalog.rst``
     - ``changes/<PR#>.spectral_leak.rst``
     - ``changes/<PR#>.srctype.rst``
     - ``changes/<PR#>.stack_refs.rst``
     - ``changes/<PR#>.straylight.rst``
     - ``changes/<PR#>.superbias.rst``
     - ``changes/<PR#>.timeconversion.rst``
     - ``changes/<PR#>.transforms.rst``
     - ``changes/<PR#>.tso_photometry.rst``
     - ``changes/<PR#>.tweakreg.rst``
     - ``changes/<PR#>.wavecorr.rst``
     - ``changes/<PR#>.wfs_combine.rst``
     - ``changes/<PR#>.wfss_contam.rst``
     - ``changes/<PR#>.white_light.rst``
     </details>
 - [ ] updated or added relevant tests
 - [ ] updated relevant documentation
 - [ ] added relevant milestone

zacharyburnett commented on .github/pull_request_template.md:
```diff
@@ -9,7 +9,83 @@ Closes #
 
 **Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
-- [ ] added entry in `CHANGES.rst` within the relevant release section
+- [ ] for a public change, added a towncrier news fragment in `changes/` <details><summary>`echo "changed something" > changes/<PR#>.<changetype>.rst`</summary>
```
**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
 - [ ] for a public change, added a towncrier news fragment in `changes/` <details><summary>`echo "changed something" > changes/<PR#>.<changetype>.rst`</summary>

     - ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
     - ``changes/<PR#>.docs.rst``
     - ``changes/<PR#>.stpipe.rst``
     - ``changes/<PR#>.datamodels.rst``
     - ``changes/<PR#>.scripts.rst``
     - ``changes/<PR#>.fits_generator.rst``
     - ``changes/<PR#>.set_telescope_pointing.rst``

     ## stage 1
     - ``changes/<PR#>.group_scale.rst``
     - ``changes/<PR#>.dq_init.rst``
     - ``changes/<PR#>.emicorr.rst``
     - ``changes/<PR#>.saturation.rst``
     - ``changes/<PR#>.ipc.rst``
     - ``changes/<PR#>.firstframe.rst``
     - ``changes/<PR#>.lastframe.rst``
     - ``changes/<PR#>.reset.rst``
     - ``changes/<PR#>.superbias.rst``
     - ``changes/<PR#>.refpix.rst``
     - ``changes/<PR#>.linearity.rst``
     - ``changes/<PR#>.rscd.rst``
     - ``changes/<PR#>.persistence.rst``
     - ``changes/<PR#>.dark_current.rst``
     - ``changes/<PR#>.charge_migration.rst``
     - ``changes/<PR#>.jump.rst``
     - ``changes/<PR#>.ramp_fitting.rst``
     - ``changes/<PR#>.gain_scale.rst``

     ## stage 2
     - ``changes/<PR#>.assign_wcs.rst``
     - ``changes/<PR#>.msaflagopen.rst``
     - ``changes/<PR#>.nsclean.rst``
     - ``changes/<PR#>.imprint.rst``
     - ``changes/<PR#>.background.rst``
     - ``changes/<PR#>.extract_2d.rst``
     - ``changes/<PR#>.master_background.rst``
     - ``changes/<PR#>.wavecorr.rst``
     - ``changes/<PR#>.srctype.rst``
     - ``changes/<PR#>.straylight.rst``
     - ``changes/<PR#>.wfss_contam.rst``
     - ``changes/<PR#>.flatfield.rst``
     - ``changes/<PR#>.fringe.rst``
     - ``changes/<PR#>.pathloss.rst``
     - ``changes/<PR#>.barshadow.rst``
     - ``changes/<PR#>.photom.rst``
     - ``changes/<PR#>.pixel_replace.rst``
     - ``changes/<PR#>.resample_spec.rst``
     - ``changes/<PR#>.residual_fringe.rst``
     - ``changes/<PR#>.cube_build.rst``
     - ``changes/<PR#>.extract_1d.rst``
     - ``changes/<PR#>.resample.rst``

     ## stage 3
     - ``changes/<PR#>.assign_mtwcs.rst``
     - ``changes/<PR#>.mrs_imatch.rst``
     - ``changes/<PR#>.tweakreg.rst``
     - ``changes/<PR#>.skymatch.rst``
     - ``changes/<PR#>.exp_to_source.rst``
     - ``changes/<PR#>.outlier_detection.rst``
     - ``changes/<PR#>.tso_photometry.rst``
     - ``changes/<PR#>.stack_refs.rst``
     - ``changes/<PR#>.align_refs.rst``
     - ``changes/<PR#>.klip.rst``
     - ``changes/<PR#>.spectral_leak.rst``
     - ``changes/<PR#>.source_catalog.rst``
     - ``changes/<PR#>.combine_1d.rst``
     - ``changes/<PR#>.ami.rst``

     ## other
     - ``changes/<PR#>.wfs_combine.rst``
     - ``changes/<PR#>.white_light.rst``
     - ``changes/<PR#>.cube_skymatch.rst``
     - ``changes/<PR#>.engdb_tools.rst``
     - ``changes/<PR#>.guider_cds.rst``
     </details>
 - [ ] updated or added relevant tests
 - [ ] updated relevant documentation
 - [ ] added relevant milestone

melanieclarke commented:
I really like this proposal in principle. I think everyone will be happier not having to de-conflict the change log all the time.  But I have a few opinions about the proposed formatting:

- I'd prefer alphabetical by step name.  It's easier to find things that way, without thinking too hard about where the step might appear in a larger context.  I do like the note about which pipelines the step appears in, though!
- The additional formatting for the subject headings and the PR links makes the file a little harder to read in plain text.  The links to PRs are useful enough in the built version that I think they're worth keeping.  But I'd prefer not to put the "``" around all the step names: they are visually distracting in both the plain text and built version of the docs.

Also, I'd like to propose that we hold off on transitioning to this system until after the next build -- development ends in just a couple weeks anyway.  That way you don't have to try to keep the fragments in line with current development, and we can start fresh with a clean change log for the next build.

tapastro commented on pyproject.toml:
```diff
@@ -286,3 +285,349 @@ lint.ignore = [
+directory = "set_telescope_pointing"
+name = '``set_telescope_pointing``'
+showcontent = true
+
```
Suggest adding a `pipeline` entry for PRs modifying the files in `jwst/pipeline`.

zacharyburnett replied:
```suggestion

# pipeline

[[tool.towncrier.type]]
directory = "pipeline"
name = 'Pipeline'
showcontent = true
```

tapastro commented on pyproject.toml:
```diff
@@ -286,3 +285,349 @@ lint.ignore = [
+directory = "background"
+name = 'background (``image2``, ``spec2``)'
+showcontent = true
+
```
Suggest adding an entry for `badpix_selfcal` step, used in spec2.

zacharyburnett replied:
```suggestion

[[tool.towncrier.type]]
directory = "badpix_selfcal"
name = 'badpix_selfcal (``spec2``)'
showcontent = true

```

zacharyburnett commented on pyproject.toml:
```diff
@@ -286,3 +285,349 @@ lint.ignore = [
+directory = "charge_migration"
+name = "charge_migration (``detector1`` NIR)"
+showcontent = true
+
```
```suggestion
[[tool.towncrier.type]]
directory = "clean_flicker_noise"
name = "clean_flicker_noise (``detector1``)"
showcontent = true

```

tapastro reviewed (changes requested):
Looks good! Just a couple of requested additions to fragment types.

melanieclarke commented:
@zacharyburnett - can we also remove the back ticks around the pipeline names in the fragment types?

melanieclarke reviewed (approved):
LGTM, when Tyler is happy.  Thanks for all the updates -- looking forward to getting this in and never de-conflicting the change log again!

tapastro reviewed (approved):
Looks good, thanks!
